File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        unordered_map<char,int>mp;
        int maxf=0;
        int r=0;
        int res=0;
        for(int i=0;i<s.size();i++){
            mp[s[i]]++;
            maxf=max(maxf,mp[s[i]]);
            while((i-r+1)-maxf>k){
                mp[s[r++]]--;
            }
            res=max(res,i-r+1);

        }
        return res;
    }
};

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int n=heights.length;
        int []nge=new int[n];
        int []pge=new int[n];
        // Stack<int> st=new Stack<>();
        Stack<Integer> st = new Stack<>();
        for(int i=0;i<n;i++){
             while(!st.empty()  && heights[i]<=heights[st.peek()]) st.pop();
             pge[i]=st.empty()?-1:st.peek();
             st.push(i);
        }
         st.clear();
        for(int i=n-1;i>=0;i--){
             while(!st.empty()  && heights[i]<=heights[st.peek()]) st.pop();
             nge[i]=st.empty()?n:st.peek();
             st.push(i);
        }
        int res=0;
        for(int i=0;i<n;i++){
           res =Math.max(res ,heights[i]*(nge[i]-pge[i]-1));
        }
        return res;
    }
}
